### PR TITLE
More tests

### PR DESCRIPTION
**Pull request by robertsLando**

### CODE REVIEW

File: test/basic.test.js [hunk 1]
```diff
@@ -23,43 +23,111 @@ test('Connect-Subscribe-Publish-Disconnect 300 clients using WS and MQTT/MQTTS p
...
+
+  var publisher = await helper.startClient('mqtt', options)
+
+  await publisher.subscribe('my/topic')
+
+  await publisher.end(true)
+
+  publisher = await helper.startClient('mqtt', options)
+
+  publisher.on('message', async function (topic, message) {
```
Comment (mcollina) [inline test/basic.test.js]: Do not pass async callbacks to event emitter. They do not do what you think they do. Nobody is doing anything if this rejects.

File: test/basic.test.js [hunk 2]
```diff
@@ -23,43 +23,111 @@ test('Connect-Subscribe-Publish-Disconnect 300 clients using WS and MQTT/MQTTS p
...
+    if (topic === 'my/topic') {
+      t.pass('Subscription has been restored')
+    }
+    await publisher.end()
```
Comment (mcollina) [inline test/basic.test.js]: This is not needed

**Review by mcollina — CHANGES_REQUESTED** [review]

File: test/basic.test.js (hunk 1 above)
Comment (robertsLando) [reply test/basic.test.js]: Just for information, how could I use an async function in a event handler?

File: test/basic.test.js (hunk 2 above)
Comment (robertsLando) [reply test/basic.test.js]: It is needed, but can be placed outside. Maybe with my latest change and sockets management I can remove this

**Comment by robertsLando** [comment]
@mcollina All tests should be ok, let me know if you would like to do more! I was thinking about making aedes process start using clusters (when persistence and mqemitter allow this), what do you think?

File: aedes.js [hunk 3]
```diff
@@ -24,6 +24,21 @@ const ports = {
...
+  sockets.add(socket)
+  socket.on('close', () => {
+    sockets.delete(socket)
+  })
+}
+
+function destroySockets () {
+  for (const s of sockets.values()) {
+    s.destroy()
+  }
```
Comment (mcollina) [inline aedes.js:41]: Why is this needed?

File: test/basic.test.js [hunk 4]
```diff
@@ -23,43 +23,214 @@ test('Connect-Subscribe-Publish-Disconnect 300 clients using WS and MQTT/MQTTS p
...
+    await publisher.publish(msg.topic, msg.payload, msg)
+    await helper.delay(100)
+
+    await Promise.all(subscribers.map(c => c.end()))
+    await publisher.end()
   }
 
-  var subscribers = []
+  await testQos(1)
+  await testQos(2)
```
Comment (mcollina) [inline test/basic.test.js]: This should be 2 tests, not one.

File: test/basic.test.js [hunk 5]
```diff
@@ -23,43 +23,214 @@ test('Connect-Subscribe-Publish-Disconnect 300 clients using WS and MQTT/MQTTS p
...
+
+  publisher.on('message', function (topic, message) {
+    if (topic === 'my/topic') {
+      t.pass('Subscription has been restored')
+    }
+  })
+
+  await publisher.publish('my/topic', 'I\'m alive', { qos: 1 })
+
+  await helper.delay(200)
```
Comment (mcollina) [inline test/basic.test.js]: Is this needed? timeouts make tests brittle

File: test/basic.test.js [hunk 6]
```diff
@@ -23,43 +23,214 @@ test('Connect-Subscribe-Publish-Disconnect 300 clients using WS and MQTT/MQTTS p
...
+          resolve(false)
+        }
+      })
+
+      subscriber.subscribe(sub, options)
+        .then(() => publisher.publish(pub, 'Test wildecards', options))
+        .catch((e) => resolve(false))
+
+      if (!result) {
+        timeout = setTimeout(resolve(true), 200)
```
Comment (mcollina) [inline test/basic.test.js]: Is this needed?

**Review by mcollina — CHANGES_REQUESTED** [review]

File: aedes.js (hunk 3 above)
Comment (robertsLando) [reply aedes.js:41]: When I try to kill aedes I will close servers but if there are existing sockets on them they will never close, this is used to keep track of existing sockets and force close them on aedes close

File: test/basic.test.js (hunk 4 above)
Comment (robertsLando) [reply test/basic.test.js]: The code is the same, in other way I have to move the function outside and create 2 tests, why should I do this?

File: test/basic.test.js (hunk 5 above)
Comment (robertsLando) [reply test/basic.test.js]: Unfortunally, I don't know why but I haven't find a way to make tap 'waiting', it always autoends when it reach the end of the function and without this it calls tearDown before the message is received

File: test/basic.test.js (hunk 6 above)
Comment (robertsLando) [reply test/basic.test.js]: When the message should NOT be received, I set a timeout, if the timeout triggers it means that is not received, is there other ways to make this check?

UPDATE: Let me try a thing, I could use t.plan for this

Comment (robertsLando) [reply test/basic.test.js]: Done in latest commit

File: test/basic.test.js (hunk 5 above)
Comment (robertsLando) [reply test/basic.test.js]: Would you like me to replace this using faketimers?

**Comment by robertsLando** [comment]
@mcollina I have added a first implementation of clusters.

There are some tests failing on both persistences. It could be caused by persistence themself and not by an error on the tests. Could you check them?

**Comment by robertsLando** [comment]
@mcollina I found that tests where not running correctly in previous version as the env var in github action must be in the same step of the command in order to work, check https://github.com/moscajs/aedes-tests/pull/3/commits/83f406046d1c429ed4be119db9551e84fed868bf. I have made some other tests locally, seems mongodb is ok, redis still have problems with wildecard tests and doesn't match empty levels (even if in latest versions of mqemitter redis and redis persistence we have added this)

**Comment by robertsLando** [comment]
@gnought Could you help with this?

**Comment by robertsLando** [comment]
@mcollina Ping

File: test/basic.test.js [hunk 7]
```diff
@@ -58,8 +57,179 @@ test('Subscribed clients receive updates', async function (t) {
...
+  await testQos(t, 1)
+})
+
+test('Subscribed clients receive updates - QoS 2', async function (t) {
+  t.plan(10, 'each client should receive a message')
+  t.tearDown(helper.closeBroker)
+
+  await helper.startBroker()
+  await testQos(t, 2)
+})
```
Comment (mcollina) [inline test/basic.test.js:77]: You should be able to include `test` inside `testQoS` and remove all the duplication

File: test/basic.test.js [hunk 8]
```diff
@@ -58,8 +57,179 @@ test('Subscribed clients receive updates', async function (t) {
...
+
+  publisher.once('message', function (topic, message) {
+    if (topic === 'my/topic') {
+      t.pass('Subscription has been restored')
+    }
+  })
+
+  await publisher.publish('my/topic', 'I\'m alive', { qos: 1 })
+
+  await helper.delay(500)
```
Comment (mcollina) [inline test/basic.test.js]: This is not needed. You can await a promise that will be resolved inside the `once('message')` handler.

File: test/basic.test.js [hunk 9]
```diff
@@ -58,8 +57,179 @@ test('Subscribed clients receive updates', async function (t) {
...
+
+  publisher = await helper.startClient()
+
+  await publisher.subscribe('test/retained/#')
+
+  publisher.on('message', function (topic, message) {
+    t.pass('Retained message received')
+  })
+
+  await helper.delay(500)
```
Comment (mcollina) [inline test/basic.test.js]: same

**Review by mcollina — CHANGES_REQUESTED** [review]

File: test/basic.test.js (hunk 7 above)
Comment (robertsLando) [reply test/basic.test.js:77]: I already did this, isn't it good?

Comment (mcollina) [reply test/basic.test.js:77]: You have two almost identical tests, one on line 66 and one on line 76. I would like this duplication to be removed by moving all content of those into the testQoS function.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by robertsLando** [comment]
@mcollina You are LGTM but tests seems not ok. I think this is due to bugs on persistences?

**Comment by mcollina** [comment]
The persistences should likely be fixed . I'm also ok to exclude some combination and open a tracking issue for the fixes.

**Comment by robertsLando** [comment]
@mcollina The problem is in mqemitter-redis that is not catching empty_levels wildecards. Will open an issue there so
